saleor/discount/utils.py

```diff
@@ -1011,16 +1012,53 @@ def fetch_promotion_rules_for_checkout(
 def _get_rule_to_channel_ids_map(rules: QuerySet):
     rule_to_channel_ids_map = defaultdict(list)
     PromotionRuleChannel = PromotionRule.channels.through
-    promotion_rule_channels = PromotionRuleChannel.objects.filter(
-        Exists(rules.filter(id=OuterRef("promotionrule_id")))
-    )
+    promotion_rule_channels = PromotionRuleChannel.objects.using(
+        settings.DATABASE_CONNECTION_REPLICA_NAME
+    ).filter(Exists(rules.filter(id=OuterRef("promotionrule_id"))))
     for promotion_rule_channel in promotion_rule_channels:
         rule_id = promotion_rule_channel.promotionrule_id
         channel_id = promotion_rule_channel.channel_id
         rule_to_channel_ids_map[rule_id].append(channel_id)
     return rule_to_channel_ids_map
 
 
+def get_channel_to_products_map_from_rules(
```

Comment:
IMO, it's hard to understand what the function is doing from the name.
As I see it must be called before `mark_products_as_dirty`. So maybe both methods should be mixed? 
Right now you need to prepare specific mapping for `mark_products_as_dirty` but maybe it could be done inside the method? WDYT?